### ISSUE TITLE
Check if point is inside mesh

### ISSUE THREAD

**Issue by sibowi** [facebookresearch/pytorch3d#343]
# Feature
Method to check if a point is located inside a mesh.

# Motivation
When working with meshes in general this is a very common problem to tackle. It would be great to have method for this in PyTorch. Making an implementation for the Meshes-class in the PyTorch3d library seems ideal.

**Comment by bottler** [comment]
When you say "inside", do you mean totally surrounded by faces in a watertight way such that there is no path to infinity from the point without crossing a face (like a generalised winding number) or something weaker like every straight ray from the point hits a face, or most rays hit a face?

**Comment by sibowi** [comment]
Yes, exactly. I mean totally surrounded by faces in a watertight way such that there is no path to infinity from the point without crossing a face (like a generalised winding number).

**Comment by gkioxari** [comment]
@sibowi 
This is a great feature request and we will add it to our list. 

Note that while we don't have the function ready-to-use we have all the components so that you can easily implement this yourself if you need it faster. We have all the functionality to compute point to triangle distances in 3D and all you need to check is that a ray starting from the point of interest intersects with a watertight mesh an odd number of times.

**Comment by sibowi** [comment]
@gkioxari 
Thanks, that sounds great!

And thanks for the pointers. I'm on it.

**Comment by ruoyuwangeel4930** [comment]
@sibowi 
Hi sibowi,

Could you please share the experience of implementing this feature? Thank you very much!

**Comment by sibowi** [comment]
Hi @ruoyuwangeel4930 

**I did make an implementation**, but haven't had the time to write it up for a proper integration with the pytorch3d library (I'm sorry!).

You can find my **code here**: https://github.com/sibowi/mesh_utils

The code is speed-optimized for elongated meshes. I.e., if your mesh elongated along the z-direction, the mesh2voxel-functionality will run significantly faster. If that is not the case it's still not too bad. This is explained a bit more in the [**tutorial**](https://github.com/sibowi/mesh_utils/blob/main/notebooks/0_mesh2voxel_tutorial.ipynb).

I validated the method on [the two example meshes](https://github.com/sibowi/mesh_utils/tree/main/resources/meshes) (also seen in [these figures](https://github.com/sibowi/mesh_utils/tree/main/figures)). And a colleague of mine has validated it on the meshes she is working with. No issues have come up (yet).

I hope this will be useful for you.

**Comment by dendenxu** [comment]
Hi, @sibowi I implemented a batched version of your [check_ray_triangle_intersection-function](https://github.com/sibowi/mesh_utils/blob/91462b0bf44bd70c79ae5a3b0453a1f0615c31dd/src/mesh_to_voxels.py#L10), hope it can help : ] !

- [Gist of the implementation](https://gist.github.com/dendenxu/ee5008acb5607195582e7983a384e644#file-moller_trumbore-py-L27).
- [Using it to check points' occupancy values](https://gist.github.com/dendenxu/ee5008acb5607195582e7983a384e644#file-moller_trumbore-py-L8)

It's not well optimized but I think it should be enough to get the job done!

**Comment by heathentw** [comment]
@gkioxari This feature still not available right......
